contracts/Crowdfunding.sol

```diff
@@ -5,99 +5,220 @@ pragma solidity ^0.8.0;
+        user2Campaigns[msg.sender].push(_campaignId);
+        campaigns[_campaignId].contributors.push(msg.sender);
     }
 
-    function getCampaigns() public view returns(Campaign[] memory) {
-        Campaign[] memory allCampaigns = new Campaign[](numCampaigns);
+    // Function to close a campaign
+    function closeCampaign(
+        uint256 _campaignId
+    ) external onlyCreator(_campaignId) {
+        require(
+            campaigns[_campaignId].state == State.OPEN,
+            "Campaign is not open"
+        );
+        require(
+            block.timestamp < campaigns[_campaignId].deadline,
+            "Campaign deadline has passed"
+        );
 
-        for (uint256 i = 0; i < numCampaigns; i++) {
-            allCampaigns[i] = campaigns[i];
+        // Check if the goal is met
+        if (
+            campaigns[_campaignId].raisedAmount >= campaigns[_campaignId].goal
+        ) {
+            // Release funds to campaign creator
+            withdraw(_campaignId);
+        } else {
+            refund(_campaignId);
         }
 
-        return allCampaigns;
+        campaigns[_campaignId].state = State.CANCEL;
+
+        //emit action
     }
 
-    function getCampaigns(uint256 _page, uint256 _pageSize) public view returns(Campaign[] memory) {
-        uint256 startIndex = _page * _pageSize;
-        require(startIndex < numCampaigns, "Invalid page number");
+    function payTo(address to, uint256 amount) internal {
+        (bool success, ) = payable(to).call{value: amount}("");
+        require(success);
+    }
 
-        uint256 endIndex = (startIndex + _pageSize) > numCampaigns ? numCampaigns : (startIndex + _pageSize);
-        Campaign[] memory campaignPage = new Campaign[](endIndex - startIndex);
+    function withdraw(uint256 _campaignId) internal {
+        address to = campaigns[_campaignId].creator;
+        uint256 amount = campaigns[_campaignId].raisedAmount;
+        payTo(to, amount);
+        //emit action
+    }
 
-        for (uint256 i = startIndex; i < endIndex; i++) {
-            campaignPage[i - startIndex] = campaigns[i];
-        }
+    function refund(uint256 _campaignId) internal {
+        // an internal function perform refunds to all the contributors of campaign _campaignId
+    }
 
-        return campaignPage;
+    function requestWithdraw(uint256 _campaignId) external {
+        // an external function call by users to request withdraw
+        // require: state is not CANCEL/CLOSE/SUCCESS; the deadline is passed and raised enough money; the msg.sender is project creator
+        // the campaign will be marked as SUCCESS after requestWithdraw
     }
 
-    // Function to create a new crowdfunding campaign
-    function createCampaign(uint256 _goal, uint256 _deadline) external {
-        require(_goal > 0, "Goal must be greater than 0");
-        require(_deadline > block.timestamp, "Deadline must be in the future");
+    function requestRefund(uint256 _campaignId) external {
```

Comment:
I'm thinking, should we add a modifier that checks if the deadline has passed, and automatically sets the campaign status to closed and refund to contributors if goal is not met?

Reply:
yea modifier is a good idea